modules/oe_content_sub_entity/src/Entity/SubEntityBase.php

```diff
@@ -28,7 +30,13 @@ abstract class SubEntityBase extends ContentEntityBase implements SubEntityInter
    */
   public function label() {
     // Return the entity's bundle label by default.
-    return $this->get('type')->entity->label();
+    $label = $this->get('type')->entity->label();
+
+    $event = new SubEntityEvent($this);
+    $event->setLabel($label);
+    $this->eventDispatcher()->dispatch(SubEntityEvent::LABEL_FORMATION, $event);
+
+    return $event->getLabel();
```

Comment:
Let's cache label to avoid possible multiple invocation of events:
```
protected $label;

public function label() {
  if (empty($this->label) {
    $event = new SubEntityEvent($this);
    $this->eventDispatcher()->dispatch(SubEntityEvent::LABEL_FORMATION, $event);
    $this->label = $event->getLabel();
  }
  return $this->label;
}
```

Reply:
Unfortunately, it will not work because many entity types depend on field values which can be changed.
I did checking `label()` methods implementation of many entity types and didn't find any usage of cache in any way.
Do you think it is worth to apply anyway caching inside label() method?

Reply:
Without caching event will be triggered every time and it will be calculate over and over again.

Reply:
But might be you are right - some dependencies can be changed. Might be we need to have a way to clear the label?

Reply:
As I can see currently we don't use some external data and in fact, we will work with already cached data.

Reply:
I would suggest to avoid triggering event every time.

```
public function label($update = FALSE) {
  if (empty($this->label) || $update) {
    $event = new SubEntityEvent($this);
    $this->eventDispatcher()->dispatch(SubEntityEvent::LABEL_FORMATION, $event);
    $this->label = $event->getLabel();
  }
  return $this->label;
}
```